src/qibo/parameter.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+import sympy as sp
+
+from qibo.config import raise_error
+
+
+class Parameter:
+    """Object which allows for variational gate parameters. Several trainable parameters
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter. All possible analytical derivatives of the lambda function are
+    calculated at the object initialisation using Sympy.
+
+    Example::
+
+        from qibo.parameter import Parameter
+        param = Parameter(
+                lambda x, th1, th2, th3: x**2 * th1 + th2 * th3**2,
+                features=[7.0],
+                trainable=[1.5, 2.0, 3.0],
+            )
+
+        partial_derivative = param.get_partial_derivative(3)
+
+        param.update_parameters(trainable=[15.0, 10.0, 7.0], feature=[5.0])
+        param_value = param()
+
+
+    Args:
+        func (function): lambda function which builds the gate parameter. If both features and trainable parameters
+        compose the function, it must be passed by first providing the features and then the parameters, as
+        described in the code example above.
+        trainable (list or np.ndarray): array with initial trainable parameters theta
+        features (list or np.ndarray): array containing possible input features x
+        nofeatures (bool): flag to explicitly ban the updating of the features. This simplifies the task of updating Parameter objects simultaneously when some have embedded features and some do not.
+    """
+
+    def __init__(self, func, trainable=None, features=None):
+        self._trainable = trainable
+        self._features = features
+
+        if self.nfeat + self.nparams != func.__code__.co_argcount:
+            raise_error(
+                TypeError,
+                f"{self.nfeat + self.nparams} parameters are provided, but {func.__code__.co_argcount} are required, please initialize features and trainable according to the defined function.",
+            )
+        # lambda function
+        self.lambdaf = func
+
+        # calculate derivatives
+        # maybe here use JAX ?
+        self.derivatives = self._calculate_derivatives()
+
+    def __call__(self):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        return self._apply_func(self.lambdaf)
+
+    @property
+    def nparams(self):
+        """Returns the number of trainable parameters"""
+        try:
+            return len(self._trainable)
+        except TypeError:
+            return 0
+
+    @property
+    def nfeat(self):
+        """Returns the number of features"""
+        return len(self._features) if isinstance(self._features, list) else 0
+
+    @property
+    def ncomponents(self):
+        """Returns the number of elements which compose the Parameter"""
+        return self.nparams + self.nfeat
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
```

Comment:
I'm not sure I understand why you need a function to call a function (i.e. one already exists, it is the application and in Python is called `__call__`).

Maybe you could just make a function to construct the parameters, but it would be better if it did not depend on `function`, like just the `else` branch of the first `if`.
Maybe it is feasible if you just define `self._trainable` to be an empty list instead of `None` (because I guess in that case you want the derivatives wrt the features, otherwise if there are no features nor trainables I do not understand why someone would use the `Parameter` class).